smu/geometry/smu_molecule.py

```diff
@@ -0,0 +1,172 @@
+"""Class that is responsible for building and assessing proposed
+   bonding patterns."""
+
+import operator
+
+from typing import Callable, Dict, List, Optional, Tuple
+
+import numpy as np
+
+from smu import dataset_pb2
+from smu.parser import smu_utils_lib
+
+
+class MatchingParameters:
+  """A class to specify optional matching parameters for SmuMolecule.place_bonds."""
+
+  def __init__(self):
+    self._must_match_all_bonds: bool = True
+
+  @property
+  def must_match_all_bonds(self) -> bool:
+    return self._must_match_all_bonds
+
+  @must_match_all_bonds.setter
+  def must_match_all_bonds(self, value: bool):
+    self._must_match_all_bonds = value
+
+
+def add_bond(a1: int, a2: int, btype: int, destination: dataset_pb2.BondTopology) -> None:
+  """Add a new Bond to `destination`.
+
+  Args:
+    a1: atom
+    a2: atom
+    btype: bond type.
+
+  """
+  destination.bonds.append(
+      dataset_pb2.BondTopology.Bond(atom_a=a1,
+                                    atom_b=a2,
+                                    bond_type=smu_utils_lib.INTEGER_TO_BOND_TYPE[btype]))
+
+
+class SmuMolecule:
+  """Holds information about partially built molecules"""
+
+  def __init__(self, hydrogens_attached: dataset_pb2.BondTopology,
+               bonds_to_scores: Dict[Tuple[int, int],
+                                     np.array], matching_parameters: MatchingParameters):
+    """Class to perform bonding assessments.
+    Args:
+      hydrogens_attached: a BondTopology that has all atoms, and the bonds
+        associated with the Hydrogen atoms.
+      bonds_to_scores: A dict that maps tuples of pairs of atoms, to a
+        numpy array of scores [0,3], for each possible bond type.
+      matching_parameters: contains possible optional behaviour modifiers.
+    Returns:
+    """
+    self._starting_bond_topology = hydrogens_attached
+    natoms = len(hydrogens_attached.atoms)
+
+    # For each atom, the maximum number of bonds that can be attached.
+    self._max_bonds = np.zeros(natoms, dtype=np.int32)
+    for i in range(0, natoms):
```

Comment:
list comprehensions are your friends :-)
self._max_bonds = np.array([smu_utils_lib.ATOM_TYPE_TO_MAX_BONDS[a] for a in hydrogens_attached.atoms])

Reply:
Yes, that looks much more pythonic. Nice.